server/pbenchinacan/deploy

```diff
@@ -1,26 +1,84 @@
+#     the appropriate location in the SRV_PBENCH file system.
+#   - The ownership of the files in the SRV_PBENCH file system will are set for
+#     access inside the container.
+#   - The Pbench Server container is started and run as a detached process.
+#
+# The key inputs are the following environment variables:
+#   - PB_DEPLOY_FILES:  the location (on the host) of the configuration files
+#                       for the Pbench Server and rsyslog.
+#   - SRV_PBENCH:  the location of the file system (on the host) which will be
+#                       mounted inside the container at /srv/pbench.
+#
+# Other inputs which can be provided by environment variables include:
+#   - the name and pull policy for the Pbench Server container image
+#   - the location of the Pbench Dashboard app files and the Nginx Favicon
+#   - the location of the Pbench Git checkout (used to locate the above)
+#
+# In all cases, suitable defaults are provided for any value which is undefined.
+#
 
-# Update the favicon
-podman unshare rm /srv/pbench/public_html/favicon.ico
-podman unshare cp color-square.ico /srv/pbench/public_html/favicon.ico
+#+
+# Configuration
+#-
 
-IMAGE=images.paas.redhat.com/pbench/pbench-server:main
+# Pbench Server container image
+PB_SERVER_IMAGE=${PB_SERVER_IMAGE:-"images.paas.redhat.com/pbench/pbench-server:main"}
+PB_SERVER_IMAGE_PULL_POLICY=${PB_SERVER_IMAGE_PULL_POLICY:-"newer"}
 
-# First ensure /srv/pbench is owned by the "pbench" user INSIDE the container we run.
-podman run --rm --volume /srv/pbench:/srv/pbench:Z ${IMAGE} chown -R pbench:pbench /srv/pbench
+# Locations on the host
+PB_DASHBOARD_DIR=${PB_DASHBOARD_DIR:-${GITTOP:=$(git rev-parse --show-toplevel)}/dashboard/build}
+NGINX_FAVICON=${NGINX_FAVICON:-${GITTOP:=$(git rev-parse --show-toplevel)}/server/lib/config/color-square.ico}
+PB_DEPLOY_FILES=${PB_DEPLOY_FILES:-${HOME}/Deploy}
+SRV_PBENCH=${SRV_PBENCH:-/srv/pbench}
 
-# Deploy the actual Pbench Server instance.
+#+
+# Deployment
+#-
+
+# Update the dashboard code, removing any existing code and copying in fresh.
+# We expect that the host dashboard directory already exists and contains a
+# build of the Pbench Dashboard application; however, if it is missing, we
+# deploy the Pbench Server without a Dashboard deployment.
+podman unshare rm -r ${SRV_PBENCH}/public_html/dashboard
+if [[ -d "${PB_DASHBOARD_DIR}" ]]; then
+    podman unshare cp -r ${PB_DASHBOARD_DIR} ${SRV_PBENCH}/public_html/dashboard
+else
+    echo "Warning:  deploying the Pbench Server without a Dashboard" >&2
+    podman unshare mkdir -m 0755 -p ${SRV_PBENCH}/public_html/dashboard
+fi
+
+# Update the Nginx favicon and provide some static fallback pages
+podman unshare rm ${SRV_PBENCH}/public_html/favicon.ico
+podman unshare cp ${NGINX_FAVICON} ${SRV_PBENCH}/public_html/favicon.ico
+podman unshare cp /usr/share/nginx/html/404.html /usr/share/nginx/html/50x.html \
+    ${SRV_PBENCH}/public_html/
```

Comment:
@portante, this step was being performed by the CI but _not_ for Staging.  Should we be doing this step for either or both, now?

Reply:
@portante...ping, here, too?

Reply:
Ideally this would be an idempotent operation handled by the deployment script to ensure the desired files are where they are supposed to be, only taking action if they are missing or not the same as the source.

Reply:
If they are the same, then overwriting them should cause no harm at little cost, no?  (And, it resolves the problem when they are missing or out of date without any checks or conditionalization.)

In any case, I'm going to take your answer as "yes for both".  🙂